src/Microsoft.Extensions.Configuration/ConfigurationRoot.cs

```diff
@@ -3,34 +3,88 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.Extensions.Primitives;
 
 namespace Microsoft.Extensions.Configuration
 {
-    public class ConfigurationRoot : ConfigurationBase, IConfigurationRoot
+    public class ConfigurationRoot : IConfigurationRoot
     {
+        private IList<IConfigurationProvider> _providers;
+        private ConfigurationReloadToken _reloadToken = new ConfigurationReloadToken();
+
         public ConfigurationRoot(IList<IConfigurationProvider> providers)
-            : base(providers)
         {
             if (providers == null)
             {
                 throw new ArgumentNullException(nameof(providers));
             }
+
+            _providers = providers;
         }
 
-        public override string Path
+        public string this[string key]
         {
             get
             {
-                return string.Empty;
+                foreach (var provider in _providers.Reverse())
+                {
+                    string value = null;
+
+                    if (provider.TryGet(key, out value))
+                    {
+                        return value;
+                    }
+                }
+
+                return null;
             }
+
+            set
+            {
+                if (!_providers.Any())
+                {
+                    throw new InvalidOperationException(Resources.Error_NoSources);
+                }
+
+                foreach (var provider in _providers)
+                {
+                    provider.Set(key, value);
+                }
+            }
+        }
+
+        public IEnumerable<IConfigurationSection> GetChildren() => GetChildrenAtPath(string.Empty);
+
+        public IEnumerable<IConfigurationSection> GetChildrenAtPath(string path)
+        {
+            var prefix = string.IsNullOrEmpty(path) ? "" : (path + Constants.KeyDelimiter);
+            return _providers
+                .Aggregate(Enumerable.Empty<string>(),
+                    (seed, source) => source.GetChildKeys(seed, path, Constants.KeyDelimiter))
+                .Distinct()
+                .Select(key => GetSection(prefix + key));
+        }
+
+        public IChangeToken GetReloadToken()
+        {
+            return _reloadToken;
+        }
+
+        public IConfigurationSection GetSection(string key)
+        {
+            return new ConfigurationSection(this, key);
         }
 
         public void Reload()
         {
-            foreach (var provider in Providers)
+            foreach (var provider in _providers)
             {
                 provider.Load();
             }
+            var previousReloadToken = Interlocked.Exchange(ref _reloadToken, new ConfigurationReloadToken());
+            previousReloadToken.OnReload();
```

Comment:
If the caller caches the token? The change token is triggered once and is a blown fuse after that - so there's a standard consumer pattern around that which avoids race conditions.